# **Penting**
- Jangan mengubah atau menambahkan cell text yang sudah disediakan, Anda hanya perlu mengerjakan cell code yang sudah disediakan.
- Pastikan seluruh kriteria memiliki output yang sesuai, karena jika tidak ada output dianggap tidak selesai.
- Misal, Anda menggunakan df = df.dropna() silakan gunakan df.isnull().sum() sebagai tanda sudah berhasil. Silakan sesuaikan seluruh output dengan perintah yang sudah disediakan.
- Pastikan Anda melakukan Run All sebelum mengirimkan submission untuk memastikan seluruh cell berjalan dengan baik.
- Pastikan Anda menggunakan variabel df dari awal sampai akhir dan tidak diperbolehkan mengganti nama variabel tersebut.
- Hapus simbol pagar (#) pada kode yang bertipe komentar jika Anda menerapkan kriteria tambahan
- Biarkan simbol pagar (#) jika Anda tidak menerapkan kriteria tambahan
- Pastikan Anda mengerjakan sesuai section yang sudah diberikan tanpa mengubah judul atau header yang disediakan.

# **1. Import Library**
Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [55]:
#Type your code here
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import joblib

from sklearn.preprocessing import MinMaxScaler, LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report


# **2. Memuat Dataset dari Hasil Clustering**
Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [44]:
# Gunakan dataset hasil clustering yang memiliki fitur Target
# Silakan gunakan dataset data_clustering jika tidak menerapkan Interpretasi Hasil Clustering [Advanced]
# Silakan gunakan dataset data_clustering_inverse jika menerapkan Interpretasi Hasil Clustering [Advanced]
# Lengkapi kode berikut
# ___ = pd_read_csv("___.csv")
df=pd.read_csv("data_clustering (1).csv")

In [45]:
# Tampilkan 5 baris pertama dengan function head.
display(df.head())

TransactionAmount  TransactionType  Location  Channel  CustomerAge  \
0           0.007207                1        36        0     0.838710   
1           0.195940                1        15        0     0.806452   
2           0.065680                1        23        2     0.016129   
3           0.096016                1        33        2     0.129032   
4           0.006874                0         1        2     0.430297   

   CustomerOccupation  TransactionDuration  LoginAttempts  AccountBalance  \
0                   0             0.244828            0.0        0.336832   
1                   0             0.451724            0.0        0.918055   
2                   3             0.158621            0.0        0.068637   
3                   3             0.051724            0.0        0.569198   
4                   3             0.648276            0.0        0.492591   

   Target  
0       0  
1       1  
2       3  
3       0  
4       2

# **3. Data Splitting**
Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [46]:
# Menggunakan train_test_split() untuk melakukan pembagian dataset.
X = df.drop("Target", axis=1)
y = df["Target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# **4. Membangun Model Klasifikasi**
Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Menggunakan algoritma klasifikasi yaitu Decision Tree.
2. Latih model menggunakan data yang sudah dipisah.

In [47]:
# Buatlah model klasifikasi menggunakan Decision Tree
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

y_pred = dt_model.predict(X_test)

In [48]:
# Menyimpan Model
# import joblib
# joblib.dump(model, 'decision_tree_model.h5')
y_pred_dt = dt_model.predict(X_test)
print("Decision Tree Classifier Report:")
print(classification_report(y_test, y_pred_dt))


joblib.dump(dt_model, 'decision_tree_model.h5')

Decision Tree Classifier Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       121
           1       1.00      1.00      1.00       131
           2       1.00      1.00      1.00       125
           3       1.00      1.00      1.00       131

    accuracy                           1.00       508
   macro avg       1.00      1.00      1.00       508
weighted avg       1.00      1.00      1.00       508



['decision_tree_model.h5']

# **5. Memenuhi Kriteria Skilled dan Advanced dalam Membangun Model Klasifikasi**



**Biarkan kosong jika tidak menerapkan kriteria skilled atau advanced**

In [49]:
# Melatih model menggunakan algoritma klasifikasi scikit-learn selain Decision Tree.
log_model = LogisticRegression(max_iter=1000, random_state=42)
log_model.fit(X_train, y_train)
log_pred = log_model.predict(X_test)

#Random Forest
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
rf_pred = rf_model.predict(X_test)

In [50]:
# Menampilkan hasil evaluasi akurasi, presisi, recall, dan F1-Score pada seluruh algoritma yang sudah dibuat.
def evaluate_model(name, y_true, y_pred):
    print(f"Hasil Evaluasi Model: {name}")
    print("Akurasi :", accuracy_score(y_true, y_pred))
    print("Presisi :", precision_score(y_true, y_pred, average='weighted'))
    print("Recall  :", recall_score(y_true, y_pred, average='weighted'))
    print("F1-Score:", f1_score(y_true, y_pred, average='weighted'))
    print("-" * 40)

evaluate_model("Decision Tree", y_test, y_pred)

evaluate_model("Random Forest", y_test, rf_pred)

evaluate_model("Logistic Regression", y_test, log_pred)

Hasil Evaluasi Model: Decision Tree
Akurasi : 1.0
Presisi : 1.0
Recall  : 1.0
F1-Score: 1.0
----------------------------------------
Hasil Evaluasi Model: Random Forest
Akurasi : 1.0
Presisi : 1.0
Recall  : 1.0
F1-Score: 1.0
----------------------------------------
Hasil Evaluasi Model: Logistic Regression
Akurasi : 1.0
Presisi : 1.0
Recall  : 1.0
F1-Score: 1.0
----------------------------------------


In [51]:
# Menyimpan Model Selain Decision Tree
# Model ini bisa lebih dari satu
# import joblib
# joblib.dump(___, 'explore_<Nama Algoritma>_classification.h5')
joblib.dump(rf_model, 'explore_random_forest_classification.h5')
joblib.dump(log_model, 'explore_logistic_regression_classification.h5')

['explore_logistic_regression_classification.h5']

Hyperparameter Tuning Model

Pilih salah satu algoritma yang ingin Anda tuning

In [52]:
# Lakukan Hyperparameter Tuning dan Latih ulang.
# Lakukan dalam satu cell ini saja.

rf = RandomForestClassifier(random_state=42)

param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

grid_search = GridSearchCV(
    estimator=rf,
    param_grid=param_grid,
    cv=5,
    scoring='accuracy',
    n_jobs=-1,
    verbose=2
)

grid_search.fit(X_train, y_train)

best_rf = grid_search.best_estimator_
best_rf.fit(X_train, y_train)
y_pred_best = best_rf.predict(X_test)

Fitting 5 folds for each of 108 candidates, totalling 540 fits


In [53]:
# Menampilkan hasil evaluasi akurasi, presisi, recall, dan F1-Score pada algoritma yang sudah dituning.
y_pred_best = best_rf.predict(X_test)
print("\n=== Tuned Random Forest Evaluation ===")
print("Accuracy:", accuracy_score(y_test, y_pred_best))
print("Precision:", precision_score(y_test, y_pred_best, average='weighted'))
print("Recall:", recall_score(y_test, y_pred_best, average='weighted'))
print("F1-Score:", f1_score(y_test, y_pred_best, average='weighted'))
print("\nClassification Report:\n", classification_report(y_test, y_pred_best))



=== Tuned Random Forest Evaluation ===
Accuracy: 1.0
Precision: 1.0
Recall: 1.0
F1-Score: 1.0

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       121
           1       1.00      1.00      1.00       131
           2       1.00      1.00      1.00       125
           3       1.00      1.00      1.00       131

    accuracy                           1.00       508
   macro avg       1.00      1.00      1.00       508
weighted avg       1.00      1.00      1.00       508



In [54]:
# Menyimpan Model hasil tuning
# import joblib
# joblib.dump(__, 'tuning_classification.h5')
joblib.dump(best_rf, 'tuning_classification.h5')

['tuning_classification.h5']

End of Code

**PENJELASAN**

**Random Forest memberikan hasil:**


*   Accuracy = 1.0 (100%)
*   Precision = 1.0 (100%)
*   Recall = 1.0 (100%)List item
*   F1-Score = 1.0 (100%)


Setiap kelas (0, 1, 2, 3) memiliki nilai precision, recall, dan f1-score = 1.00 dengan jumlah data uji (support) total 508.


---



**Penjelasan Hasil Klasifikasi**


1. **kinerja model sangat optimal**
Semua metrik evaluasi (akurasi, presisi, recall, dan f1-score) bernilai 100%. Hal ini menunjukkan bahwa model mampu mengenali seluruh kelas (cluster) tanpa kesalahan prediksi sama sekali pada data uji.

2.  **Setiap Cluster Terklasifikasi dengan Sempurna**

*   Cluster 0: Semua data berhasil diprediksi dengan benar (121 sampel).
*   Cluster 1: Tidak ada kesalahan prediksi (131 sampel).
*   Cluster 2: Seluruh data dikenali dengan tepat (125 sampel).
*   Cluster 3: Semua sampel terdeteksi sesuai kelasnya (131 sampel).


3. **Interpretasi**
Hasil tersebut mengindikasikan bahwa model dari Random Forest yang telah di-tuning sangat efektif untuk berfungsi membedakan pola dari tiap cluster yang dihasilkan sebelumnya. Model ini mampu menangkap perbedaan karakteristik transaksi setiap nasabah secara konsisten, sehingga tidak ada data uji yang mengalami kesalahan pada saat klasifikasi.


*   Dataset yang digunakan bisa terbilang cukup bersih dan telah di proses dengan baik melalui proses preprocessing dan clustering.

*   model random forest dengan parameter yang telah di-tune bisa di bilang sesuai dengan dataset ini.
*   dataset mungkin masih memiliki kompleksitas yang renda, sehingga model masih bisa untuk mempelajari pola dengan sangat baik.





